File: lib/features/activity_freediving/models/freediving_conditions_report.dart
import 'package:turbo/features/activity_fishing/api.dart' show WeatherSlice;

class FreedivingConditionsReport {
  final String activityId;
  final DateTime validAt;
  final DateTime fetchedAt;
  final WeatherSlice weather;
  final String? seaStateSummary;
  final int? score;
  final String rationale;

  const FreedivingConditionsReport({
    required this.activityId,
    required this.validAt,
    required this.fetchedAt,
    required this.weather,
    required this.seaStateSummary,
    required this.score,
    required this.rationale,
  });

  factory FreedivingConditionsReport.fromJson(Map<String, dynamic> json) =>
      FreedivingConditionsReport(
        activityId: json['activityId'] as String,
        validAt: DateTime.parse(json['validAt'] as String),
        fetchedAt: DateTime.parse(json['fetchedAt'] as String),
        weather: WeatherSlice.fromJson(json['weather'] as Map<String, dynamic>),
        seaStateSummary: json['seaStateSummary'] as String?,
        score: (json['score'] as num?)?.toInt(),
        rationale: json['rationale'] as String,
      );
}

File: apps/api/src/Activities/Turbo.Activities.Freediving.Contracts/value/FreedivingConditionsReport.cs
using System.Text.Json.Serialization;
using Turboapi.Activities.value;

namespace Turboapi.Activities.Freediving.value;

public sealed record FreedivingConditionsReport
{
    [JsonPropertyName("activityId")] public Guid ActivityId { get; init; }
    [JsonPropertyName("validAt")] public DateTimeOffset ValidAt { get; init; }
    [JsonPropertyName("fetchedAt")] public DateTimeOffset FetchedAt { get; init; }
    [JsonPropertyName("weather")] public WeatherSlice Weather { get; init; }

    /// <summary>Sea state at the spot (m³/s current, or wave-height
    /// proxy). Null until the TidesProvider lands.</summary>
    [JsonPropertyName("seaStateSummary")] public string? SeaStateSummary { get; init; }

    [JsonPropertyName("score")] public int? Score { get; init; }
    [JsonPropertyName("rationale")] public string Rationale { get; init; }

    [JsonConstructor]
    public FreedivingConditionsReport(
        Guid activityId, DateTimeOffset validAt, DateTimeOffset fetchedAt,
        WeatherSlice weather, string? seaStateSummary, int? score, string rationale)
    {
        ActivityId = activityId;
        ValidAt = validAt;
        FetchedAt = fetchedAt;
        Weather = weather;
        SeaStateSummary = seaStateSummary;
        Score = score;
        Rationale = rationale ?? throw new ArgumentNullException(nameof(rationale));
    }
}
